plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java

```diff
@@ -64,8 +67,10 @@ public Procedure get()
 
     public void rollbackToSnapshot(ConnectorSession clientSession, String schema, String table, Long snapshotId)
     {
-        SchemaTableName schemaTableName = new SchemaTableName(schema, table);
-        Table icebergTable = catalogFactory.create(clientSession.getIdentity()).loadTable(clientSession, schemaTableName);
-        icebergTable.rollback().toSnapshotId(snapshotId).commit();
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
```

Comment:
Could you please add a code comment related to the reasoning behind this change?
This approach is at the moment of this writing rather cryptical to me.

Reply:
@findepi does a great job in explaining this thing, so maybe you should talk to him ;)

Basically in the ticket mentioned by you (https://github.com/trinodb/trino/pull/12687) the check is added that makes sure that only one classloader ever creates new instances of `org.apache.hadoop.conf.Configuration` class. 
When trino is run normally(so not in tests), at the start, it initiates plugins(connectors) and for this uses `PluginClassLoader`. 
Hive, Iceberg and DeltaLake connectors happen to use that `Configuration` class during initialization so it  is `PluginClassLoader` that creates first instances of it and it is stored [here](https://github.com/trinodb/trino/blob/8cb79f08493c949f75b48fc102475ede15e331de/lib/trino-hadoop-toolkit/src/main/java/io/trino/hadoop/ConfigurationInstantiator.java#L42).
However when `rollback_to_snapshot` procedure was called its [rollbackToSnapshot](https://github.com/trinodb/trino/blob/8cb79f08493c949f75b48fc102475ede15e331de/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java#L65)
is called from a Server code not from a connector code and thus different class loader is used `AppClassLoader`.
This change forces using correct class loader, as this class as a whole is part of an iceberg connector so it is loaded/initialized by `PluginClassLoader`, so it is enough to just catch the classLoader used during initialization and use it later. 

Does that make sense? @findepi @findinpath

Reply:
Btw i will add a code comment but little shorter than this

Reply:
> Does that make sense?

tl;dr ;)

Reply:
> However when rollback_to_snapshot procedure was called its rollbackToSnapshot is called from a server code ...

Does this actually mean that all  (hive/iceberg/delta) stored procedure calls are / may be affected by this problem?

The method `methodHandle` used in the `ROLLBACK_TO_SNAPSHOT` could be probably then adjusted/ slightly duplicated so that it does (under the scenes) the trick you are doing within this class and therefore transparently 
address the problem for the other stored procedure calls that may be uncovered with #12687

```
this.classLoader = Thread.currentThread().getContextClassLoader();
```

Which class loader are we expected to deal with here? Is it the  PluginClassLoader / AppClassLoader/ something else?




> Does that make sense?

Yes, I think I have a better context of the fix now.

Reply:
> Does this actually mean that all (hive/iceberg/delta) stored procedure calls are / may be affected by this problem?

as long as they try to create new instance of Configuration class they may be affected

> The method methodHandle used in the ROLLBACK_TO_SNAPSHOT could be probably then adjusted/ slightly duplicated so that it does (under the scenes) the trick you are doing within this class and therefore transparently
address the problem for the other stored procedure calls that may be uncovered with https://github.com/trinodb/trino/pull/12687

I may be missing something but I don't see how this method could be used. It is static so it is not possible to force it to store the reference to the required class loader.  In theory we could reuse the class loader that loaded `MethodHandleUtil` from which that method comes but we would then need to make sure that each connector initialize that class - I also think that would be equally if not more hacky. 

> Which class loader are we expected to deal with here? Is it the PluginClassLoader / AppClassLoader/ something else?

We need PluginClassLoader here.